Generate Client using OpenAPI spec tool, Update Setup Guide and Quickstart

sajitha-tj opened the pull request:
## Purpose
Initial setup of the ballerina connector
## Examples

## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests
- [x] Updated the spec
- [ ] Checked native-image compatibility

sajitha-tj committed:
Generate client code using Ballerina OpenAPI tool

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
Succesfully build the project

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
Update package.md file

sajitha-tj committed:
update folder structure of docs

sajitha-tj committed:
[Automated] Update the toml files

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ballerina-platform/module-ballerinax-hubspot.crm.lists?pullRequest=2) <br/>All committers have signed the CLA.

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
[Automated] Update the toml files

tharindu-nw commented on ballerina/tests/mock_service.bal:
```diff
@@ -0,0 +1,9 @@
+        response.statusCode = 204;
+        return response;
+    }
+}
```
Shall we add a new line at the end of file.
Also applicable for the following files:
- ballerina/tests/tests.bal
- docs/spec/openapi.yaml

sajitha-tj replied:
Done. [4b5c1a2](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/4b5c1a2f17342559cb349c6dd44fa56b8f2e4da4)

sajitha-tj committed:
Add test cases for some api endpoints

sajitha-tj committed:
Update package.md

sajitha-tj committed:
Add sanitations to the OpenAPI specs

sajitha-tj committed:
Update ballerina client after sanitations of openapi spec

All of the API endpoints had the common prefix `/crm/v3/lists`. Those paths are modified to remove this prefix from the endpoints, and included in the base url. This centralizes the versioning to the base url and simplifies the API paths, making them more readable.

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
Added new test cases for other endpoints

sajitha-tj committed:
Add new line at the end of file

sajitha-tj committed:
Do sanitations on open API specs

As mentioned in the sanitations.md file, few feilds needed changes inorder to generate the connector client correctly.

sajitha-tj committed:
Add minor changes in package.md file

sajitha-tj committed:
Add test cases for more API endpoints

sajitha-tj committed:
Change client methods from resource to remote to fix ambiguity issue

Previously used resource methods caused a "resource method is ambiguous" error for the endpoint "/crm/v3/lists/folders". Inorder to resolve this error, Ballerina connector was regenerated with remote methods and this fixed the issue.

Additionally, all test cases were updated accordingly with the new generated method names, ensuring compatibility with the updated client methods.

sajitha-tj committed:
format tests.bal using bal format command

sajitha-tj committed:
Update sanitations.md file

Ballerina client is generated with remote methods to fix the ambiguity issue. The command used for generating the client using openAPI tool changed for this purpose.

Also did some minor updates on the documentation.

ThisaruGuruge commented:
Shall we update the README file?

sajitha-tj committed:
Add an example usecase of the CRM lists connector

This example demonstrates how to use the HubSpot CRM Lists Connector to create several lists which filters customer support tickets with respect to their priority level. It shows how to use multiple methods of the connector with each other effectively.

Since the example uses tickets, the OAuth key should have the consent for managing tickets.

sajitha-tj committed:
Update test cases

Changed test cases to use 'assertEquals' instead of 'assertTrue' at suitable places. These changes improve the readability and maintainability of the tests.

ThisaruGuruge commented on build-config/resources/Ballerina.toml:
Let's add the keywords in this file:

```
keywords = ["hubspot", "crm", "lists"]

sajitha-tj replied:
Done. [199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

ThisaruGuruge commented on ballerina/Package.md:
```diff
@@ -1,14 +1,189 @@
-[//]: # (TODO: Add overview mentioning the purpose of the module, supported REST API versions, and other high-level details.)
+[HubSpot](https://www.hubspot.com/) is an AI-powered customer relationship management (CRM) platform. 
+
+The ballerinax/hubspot.crm.lists offers APIs to connect and interact with the [HubSpot CRM Lists API](https://developers.hubspot.com/docs/reference/api/crm/lists) endpoints, specifically based on the [HubSpot CRM Lists API v3 OpenAPI specification](https://github.com/HubSpot/HubSpot-public-api-spec-collection/blob/main/PublicApiSpecs/CRM/Lists/Rollouts/144891/v3/lists.json)
```
```suggestion
The `ballerinax/hubspot.crm.lists` offers APIs to connect and interact with the [HubSpot CRM Lists API](https://developers.hubspot.com/docs/reference/api/crm/lists) endpoints, specifically based on the [HubSpot REST API](https://developers.hubspot.com/docs/reference/api/overview)
```

ThisaruGuruge commented on ballerina/Package.md:
```diff
@@ -1,14 +1,189 @@
+
+If you don't have a HubSpot Developer Account you can sign up to a free account [here](https://developers.hubspot.com/get-started)
+
+### Step 2 (Optional): Create a [Developer Test Account](https://developers.hubspot.com/beta-docs/getting-started/account-types#developer-test-accounts) under your account
+
+Within app developer accounts, you can create developer test accounts to test apps and integrations without affecting any real HubSpot data.
+
+**_These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts._**
```
```suggestion
### Step 2 (Optional): Create a Developer Test Account

Within app developer accounts, you can create a [developer test account](https://developers.hubspot.com/beta-docs/getting-started/account-types#developer-test-accounts) under your account to test apps and integrations without affecting any real HubSpot data.

> **Note:** These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts.
```

ThisaruGuruge commented on ballerina/Package.md:
```diff
@@ -1,14 +1,189 @@
+
+1. Go to Test Account section from the left sidebar.
+
+   <img src=../docs/setup/resources/test_acc_1.png alt="Hubspot developer portal" style="width: 70%;">
```
Please use the markdown format here. Check similar places.

```suggestion
![Hubspot developer portal](./docs/setup/resources/test_acc_1.png)
```

sajitha-tj replied:
Done. [3290d95](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/3290d952980ad9d39a0ebde71451862de5deea99)

ThisaruGuruge commented on ballerina/Package.md:
```diff
@@ -1,14 +1,189 @@
+   <img src=../docs/setup/resources/create_app_2.png alt="Selecting the Auth tab" style="width: 70%;">
+
+2. In the Scopes section, add the following scopes for your app using the "Add new scope" button.
+
+   `crm.lists.read`
+
+   `crm.lists.write`
+
+   `cms.membership.access_groups.write`
```
Use a list here.

```suggestion
    * `crm.lists.read`
    * `crm.lists.write`
    * `cms.membership.access_groups.write`
```

sajitha-tj replied:
Done. [199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

ThisaruGuruge commented on ballerina/Package.md:
```diff
@@ -1,14 +1,189 @@
+
+   <img src=../docs/setup/resources/scope_set.png alt="List of scopes" style="width: 70%;">
+
+   Scopes listed above are the mandotory scopes needed to use the HubSpot CRM Lists API. However you may need to add additional scopes based on your usecase. For example, if you are working with contacts, you may need to add `crm.objects.contacts.read` and `crm.objects.contacts.write` scopes as well.
```
We can remove this part.

````suggestion
````

sajitha-tj replied:
Done. [199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

sajitha-tj replied:
Had to add this part again because working with some endpoints needs other consent for other scopes as well. [2e03824](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/2e03824283846e1a70eca2beed35304f71548554)

ThisaruGuruge commented on ballerina/Package.md:
```diff
@@ -1,14 +1,189 @@
+
+   Scopes listed above are the mandotory scopes needed to use the HubSpot CRM Lists API. However you may need to add additional scopes based on your usecase. For example, if you are working with contacts, you may need to add `crm.objects.contacts.read` and `crm.objects.contacts.write` scopes as well.
+
+3. Add your Redirect URI in the relevant section. You can also use localhost addresses for local development purposes. Click Create App.
```
```suggestion
3. Add your Redirect URI in the relevant section. You can also use `localhost` addresses for local development purposes. Click "Create App".
```

sajitha-tj replied:
Done. [199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

ThisaruGuruge commented on ballerina/Package.md:
```diff
@@ -1,14 +1,189 @@
+
+   Replace the `<YOUR_CLIENT_ID>`, `<YOUR_REDIRECT_URI>` and `<YOUR_SCOPES>` with your specific value.
+
+    **_NOTE: If you are using a localhost redirect url, make sure to have a listner running at the relevant port before executing the next step. You can use [this gist](https://gist.github.com/lnash94/0af47bfcb7cc1e3d59e06364b3c86b59) and run it using `bal run`. Alternatively, you can use any other method to bind a listner to the port._**
```
Let's remove this part.
```suggestion
```

sajitha-tj replied:
Done. [199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

ThisaruGuruge commented on ballerina/Package.md:
````diff
@@ -1,14 +1,189 @@
+   <img src=../docs/setup/resources/install_app.png alt="Selecting the App for OAuth process" style="width: 70%;">
+
+3. A code will be displayed in the browser. Copy the code.
+
+   ```
+   Received code: na1-129d-860c-xxxx-xxxx-xxxxxxxxxxxx
+   ```
````
Remove this part.

```suggestion
```

sajitha-tj replied:
Done.[199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

ThisaruGuruge commented on ballerina/Package.md:
```diff
@@ -1,14 +1,189 @@
+        credentialBearer: oauth2:POST_BODY_BEARER
+    };
+
+    final string serviceUrl = "https://api.hubapi.com";
+
+    final crmlists:Client crmListClient = check new Client(config = {auth}, serviceUrl = serviceUrl);
```
We can omit the service URL here.

```suggestion
    final crmlists:Client crmListClient = check new Client(config = {auth});
```

sajitha-tj replied:
Done. [199344e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/199344e071e5fefbc37a599d02123f8be886df75)

ThisaruGuruge commented on ballerina/tests/tests.bal:
```diff
@@ -0,0 +1,350 @@
+    ListCreateResponse response = check hubspotClient->post_create(payload);
+    testListId = response.list.listId;
+    testListName = response.list.name;
+    test:assertTrue(response.list.name == testListName);
```
```suggestion
    test:assertEquals(response.list.name, testListName);
```

sajitha-tj replied:
Done. [ff8ee73](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/ff8ee7381e1d609c05c65ac72ef47f9c6ae38381)

ThisaruGuruge commented on ballerina/tests/tests.bal:
```diff
@@ -0,0 +1,350 @@
+    };
+    ListCreateResponse response = check hubspotClient->post_create(payload);
+    testDynamicListId = response.list.listId;
+    test:assertTrue(response.list.name == "my-test-list-dynamic");
```
Check similar places.

```suggestion
    test:assertEquals(response.list.name, "my-test-list-dynamic");
```

sajitha-tj replied:
Done [ff8ee73](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/ff8ee7381e1d609c05c65ac72ef47f9c6ae38381)

ThisaruGuruge commented on ballerina/tests/tests.bal:
```diff
@@ -0,0 +1,350 @@
+    };
+    ListSearchResponse response = check hubspotClient->postSearch_dosearch(payload);
+    test:assertTrue(response.lists.length() >= 0);
+    if (response.lists.length() > 0) {
+        test:assertTrue(response.lists[0].name.includes("test"));
+    }
```
Since the length was asserted in the above line, no need of checking it again.

```suggestion
    test:assertTrue(response.lists[0].name.includes("test"));
```

sajitha-tj replied:
Ok. [f3f6cf2](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/f3f6cf2f9b51baefc96cd7c41ee5939571c76c71)

ThisaruGuruge commented on ballerina/tests/tests.bal:
```diff
@@ -0,0 +1,350 @@
+        query: "test"
+    };
+    ListSearchResponse response = check hubspotClient->postSearch_dosearch(payload);
+    test:assertTrue(response.lists.length() >= 0);
```
Shouldn't this be greater than 0?

```suggestion
    test:assertTrue(response.lists.length() > 0);
```

sajitha-tj replied:
Yes. I fixed that here [f3f6cf2](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/f3f6cf2f9b51baefc96cd7c41ee5939571c76c71)

ThisaruGuruge commented on ballerina/tests/tests.bal:
```diff
@@ -0,0 +1,350 @@
+        name: "test-folder"
+    };
+    ListFolderCreateResponse response = check hubspotClient->postFolders_create(payload);
+    test:assertTrue(response.folder.name == "test-folder");
```
Check similar places.

```suggestion
    test:assertEquals(response.folder.name, "test-folder");
```

sajitha-tj replied:
Done. [ff8ee73](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/ff8ee7381e1d609c05c65ac72ef47f9c6ae38381)

ThisaruGuruge commented on examples/customer_support_tickets_manager/Ballerina.toml:
```diff
@@ -0,0 +1,14 @@
+[build-options]
+observabilityIncluded = true
+
+[[dependency]]
+org="ballerinax"
+name="hubspot.crm.lists"
+version="0.1.0"
+repository="local"
```
Let's remove the local dependency
```suggestion
```

sajitha-tj replied:
Done. [d20ad6c](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/d20ad6ce821d55e9222991248d0bee56907fe632)

ThisaruGuruge commented on examples/customer_support_tickets_manager/main.bal:
```diff
@@ -0,0 +1,190 @@
+    if postFolderCreateResponse is error {
+        io:println("[!] Error occurred while creating the folder: ", postFolderCreateResponse);
+        return;
+    }
```
```suggestion
    hubspotcrmLists:ListFolderCreateResponse postFolderCreateResponse = check hubspotClient->postFolders_create(folderCreateReqPayload);
```

sajitha-tj replied:
Done [d20ad6c](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/d20ad6ce821d55e9222991248d0bee56907fe632)

ThisaruGuruge commented on examples/customer_support_tickets_manager/main.bal:
```diff
@@ -0,0 +1,190 @@
+    io:println("[+] Folder created successfully: ", folder.name);
+
+    // Generating lists for seperate priority levels
+    hubspotcrmLists:PublicEnumerationPropertyOperation filterOperation ={
```
Check the formatting

```suggestion
    hubspotcrmLists:PublicEnumerationPropertyOperation filterOperation = {
```

sajitha-tj replied:
Done. [7c977a2](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/7c977a28eba5e7ca58346fdb87e260138a913fe2)

ThisaruGuruge commented on examples/customer_support_tickets_manager/main.bal:
```diff
@@ -0,0 +1,190 @@
+
+    // Generating lists for seperate priority levels
+    hubspotcrmLists:PublicEnumerationPropertyOperation filterOperation ={
+    includeObjectsWithNoValueSet: false,
+    values: ["HIGH"],
+    operationType: "ENUMERATION",
+    operator: "IS_ANY_OF"
```
```suggestion
        includeObjectsWithNoValueSet: false,
        values: ["HIGH"],
        operationType: "ENUMERATION",
        operator: "IS_ANY_OF"
```

sajitha-tj replied:
fixed. [7c977a2](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/7c977a28eba5e7ca58346fdb87e260138a913fe2)

ThisaruGuruge commented on examples/customer_support_tickets_manager/main.bal:
```diff
@@ -0,0 +1,190 @@
+    hubspotcrmLists:PublicObjectList mediumPriorityList;
+    hubspotcrmLists:PublicObjectList lowPriorityList;
+
+    if highPriorityListCheckResponse is error{
+        hubspotcrmLists:ListCreateResponse|error highPriorityListResponse = hubspotClient->post_create(payload=payloadForLists);
+        if highPriorityListResponse is error {
+            io:println("[!] Error occurred while creating the high priority list: ", highPriorityListResponse);
+            return;
+        }
+        highPriorityList = highPriorityListResponse.list;
+        io:println("[+] High priority list created successfully: ", highPriorityList.name);
+    }else{
+        highPriorityList = highPriorityListCheckResponse.list;
+        io:println("[+] High priority list already exists");
+    }
```
No need to print the error each time.

```suggestion
    if highPriorityListCheckResponse is error{
        hubspotcrmLists:ListCreateResponse highPriorityListResponse = check hubspotClient->post_create(payload=payloadForLists);
        highPriorityList = highPriorityListResponse.list;
        io:println("[+] High priority list created successfully: ", highPriorityList.name);
    }else{
        highPriorityList = highPriorityListCheckResponse.list;
        io:println("[+] High priority list already exists");
    }
```

sajitha-tj replied:
Done [d20ad6c](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/d20ad6ce821d55e9222991248d0bee56907fe632)

ThisaruGuruge commented on examples/customer_support_tickets_manager/main.bal:
```diff
@@ -0,0 +1,190 @@
+        }
+        mediumPriorityList = mediumPriorityListResponse.list;
+        io:println("[+] Medium priority list created successfully: ", mediumPriorityList.name);
+    }else {
```
Check formatting
```suggestion
    } else {
```

sajitha-tj replied:
Done. [7c977a2](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/7c977a28eba5e7ca58346fdb87e260138a913fe2)

ThisaruGuruge commented on examples/customer_support_tickets_manager/main.bal:
```diff
@@ -0,0 +1,190 @@
+        io:println("    - ", member.recordId);
+    }
+
+    io:println();
```
```suggestion
```

sajitha-tj replied:
Done [d20ad6c](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/d20ad6ce821d55e9222991248d0bee56907fe632)

sajitha-tj committed:
Format example code with bal format command

sajitha-tj committed:
Update ballerina/Package.md

Co-authored-by: Thisaru Guruge <thisaru@wso2.com>

sajitha-tj committed:
Update ballerina/Package.md

Co-authored-by: Thisaru Guruge <thisaru@wso2.com>

sajitha-tj committed:
Merge branch 'main' of https://github.com/sajitha-tj/module-ballerinax-hubspot.crm.lists

sajitha-tj committed:
Update package.md file

sajitha-tj committed:
Update Package.md file

sajitha-tj committed:
Update Package.md file

For working with some endpoints, OAuth key needs consent for some additional scopes. Without the consent for those scopes, we cannot create or read Lists of that object type. This is explained in the Package.md file.

sajitha-tj committed:
Update tests with better assertions

Since the length is asserted in the previos line, the if statement is removed.

sajitha-tj committed:
Update examples code and remove local dependency from Ballerina.toml

There were some duplicate error checks in the code which were removed and added 'check' statements as suggested in the PR.

sajitha-tj committed:
Update Readme file for examples

sajitha-tj committed:
Add another example usecase of the connector

'Leads Tracker' Example uses Ballerina HubSpot CRM Lists connector to create CRM lists and add records to those lists from the CLI. Example uses 'MANUAL' type lists for demonstration purposes.

sajitha-tj committed:
Update readme file for the new example

Readme file is updated with the details of the new example. Moreover, leads_tracker.md file was added which contains details about the example and how to configure and run it.

sajitha-tj committed:
Format leads_tracker example with bal format

sajitha-tj committed:
Remove local dependencies from Ballerina.toml file

sajitha-tj committed:
Update README.md and some other documentations

README.md and the module.md files were updated with necessary details about the package. Also did some minor changes on other documentations as well.

sajitha-tj committed:
Update sanitations.md

sajitha-tj committed:
Replace relative paths of images in docs with absolute paths

sajitha-tj committed:
Add a mock service for running mock tests

A mock service which simulates some of HubSpot CRM API responses was added to allow local testing without hitting the actual HubSpot server. Tests in the tests.bal file were grouped based on the ability to running them with mock server.

sajitha-tj committed:
Fix spelling errors on documentations

ThisaruGuruge commented on README.md:
```diff
@@ -8,21 +8,187 @@
+        credentialBearer: oauth2:POST_BODY_BEARER
+    };
+
+    final crmlists:Client crmListClient = check new Client(config = {auth});
```
```suggestion
    final crmlists:Client crmListClient = check new ({auth});
```

ThisaruGuruge commented on README.md:
````diff
@@ -8,21 +8,187 @@
+Import the `hubspot.crm.lists` module and `oauth2` module.
+
+```ballerina
+import ballerinax/hubspot.crm.lists as crmlists;
````
Let's use `hslists` alias here.

```suggestion
import ballerinax/hubspot.crm.lists as hslists;
```

sajitha-tj replied:
Done

ThisaruGuruge commented on ballerina/Module.md:
````diff
@@ -1,17 +1,183 @@
+Import the `hubspot.crm.lists` module and `oauth2` module.
+
+```ballerina
+import ballerinax/hubspot.crm.lists as crmlists;
````
Let's change this to `hslists`. Please check similar places.

```suggestion
import ballerinax/hubspot.crm.lists as hslists;
```

sajitha-tj replied:
Done

ThisaruGuruge commented on ballerina/Module.md:
```diff
@@ -1,17 +1,183 @@
+        credentialBearer: oauth2:POST_BODY_BEARER
+    };
+
+    final crmlists:Client crmListClient = check new Client(config = {auth});
```
```suggestion
    final crmlists:Client crmListClient = check new ({auth});
```

sajitha-tj replied:
Done

ThisaruGuruge commented on ballerina/Package.md:
Do the changes suggested on `Module.md` file in here as well.

sajitha-tj replied:
Done. [a051732](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/a051732b30bd6d54faaed900cd46b0a4efa30bd4)

ThisaruGuruge commented on ballerina/tests/tests.bal:
```diff
@@ -0,0 +1,403 @@
+import ballerina/oauth2;
+import ballerina/test;
+
+configurable boolean isLiveServer = ?; // Set this to true to run live tests with hubspot API
```
```suggestion
configurable boolean isLiveServer = false; // Set this to true to run live tests with hubspot API
```

ThisaruGuruge commented on ballerina/tests/tests.bal:
```diff
@@ -0,0 +1,403 @@
+        hubspotClient = check new Client(config = {auth});
+    } else {
+        hubspotClient = check new Client(config = {auth}, serviceUrl = "http://localhost:9090");
+    }
```
Avoid using parenthesis around the `if` conditions, please check similar places.

```suggestion
    if isLiveServer {
        hubspotClient = check new ({auth});
    } else {
        hubspotClient = check new ({auth}, serviceUrl = "http://localhost:9090");
    }
```

sajitha-tj replied:
Done. [a1016b3](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/a1016b385e9232c7a5680b5ad7ec346104867022)

ThisaruGuruge commented on ballerina/tests/tests.bal:
Good job!

sajitha-tj replied:
Thank you!

ThisaruGuruge commented on build-config/resources/Ballerina.toml:
Add the keywords in this file, and build the connector locally to apply the keywords in the `Ballerina.toml` file in the `ballerina` directory. 

```toml
keywords = ["hubspot", "crm"]
```

sajitha-tj replied:
Done. [290d00e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/290d00eb930bb6f012c3674468898dc0b3f652f5)

ThisaruGuruge commented on examples/customer_support_tickets_manager/.devcontainer.json:
Remove this file.

sajitha-tj replied:
Done. [290d00e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/290d00eb930bb6f012c3674468898dc0b3f652f5)

ThisaruGuruge commented on examples/customer_support_tickets_manager/.gitignore:
Remove this file.

sajitha-tj replied:
Done. [290d00e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/290d00eb930bb6f012c3674468898dc0b3f652f5)

ThisaruGuruge commented on examples/customer_support_tickets_manager/main.bal:
```diff
@@ -0,0 +1,149 @@
+
+import ballerina/io;
+import ballerina/oauth2;
+import ballerinax/hubspot.crm.lists as hubspotcrmLists;
```
Let's use `hslists` as the alias.

```suggestion
import ballerinax/hubspot.crm.lists as hslists;
```

sajitha-tj replied:
Done. [290d00e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/290d00eb930bb6f012c3674468898dc0b3f652f5)

ThisaruGuruge commented on examples/customer_support_tickets_manager/main.bal:
```diff
@@ -0,0 +1,149 @@
+    credentialBearer: oauth2:POST_BODY_BEARER
+};
+
+final hubspotcrmLists:Client hubspotClient = check new hubspotcrmLists:Client(config = {auth});
```
```suggestion
final hubspotcrmLists:Client hubspotClient = check new ({auth});
```

sajitha-tj replied:
Done. [290d00e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/290d00eb930bb6f012c3674468898dc0b3f652f5)

ThisaruGuruge commented on examples/customer_support_tickets_manager/main.bal:
```diff
@@ -0,0 +1,149 @@
+    // Move the created lists to the folder
+    _ = check hubspotClient->putFoldersMoveList_movelist(payload = {listId: highPriorityList.listId, newFolderId: folder.id.toString()});
+    _ = check hubspotClient->putFoldersMoveList_movelist(payload = {listId: mediumPriorityList.listId, newFolderId: folder.id.toString()});
+    _ = check hubspotClient->putFoldersMoveList_movelist(payload = {listId: lowPriorityList.listId, newFolderId: folder.id.toString()});
```
Don't have to pass the parameter name here. Check if this fails. If it doesn't, check and change every similar occurrence. 

```suggestion
    _ = check hubspotClient->putFoldersMoveList_movelist({listId: highPriorityList.listId, newFolderId: folder.id.toString()});
    _ = check hubspotClient->putFoldersMoveList_movelist({listId: mediumPriorityList.listId, newFolderId: folder.id.toString()});
    _ = check hubspotClient->putFoldersMoveList_movelist({listId: lowPriorityList.listId, newFolderId: folder.id.toString()});
```

sajitha-tj replied:
Done. [290d00e](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/290d00eb930bb6f012c3674468898dc0b3f652f5)

ThisaruGuruge commented on examples/leads_tracker/.devcontainer.json:
remove this file.

sajitha-tj replied:
Done. [bdf091a](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/bdf091a0a9b82ddc6880244d9c229650fa369357)

ThisaruGuruge commented on examples/leads_tracker/.gitignore:
Remove this file.

sajitha-tj replied:
Done. [bdf091a](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/bdf091a0a9b82ddc6880244d9c229650fa369357)

ThisaruGuruge commented on examples/leads_tracker/main.bal:
```diff
@@ -0,0 +1,111 @@
+
+import ballerina/io;
+import ballerina/oauth2;
+import ballerinax/hubspot.crm.lists as hubspotcrmLists;
```
```suggestion
import ballerinax/hubspot.crm.lists as hslists;
```

sajitha-tj replied:
Done. [bdf091a](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/bdf091a0a9b82ddc6880244d9c229650fa369357)

ThisaruGuruge commented on examples/leads_tracker/main.bal:
```diff
@@ -0,0 +1,111 @@
+    credentialBearer: oauth2:POST_BODY_BEARER
+};
+
+final hubspotcrmLists:Client hubspotClient = check new hubspotcrmLists:Client(config = {auth});
```
```suggestion
final hubspotcrmLists:Client hubspotClient = check new ({auth});
```

sajitha-tj replied:
Done. [bdf091a](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/bdf091a0a9b82ddc6880244d9c229650fa369357)

ThisaruGuruge commented on examples/leads_tracker/main.bal:
```diff
@@ -0,0 +1,111 @@
+        leadsIDs.push(leadID);
+    }
+    return leadsIDs;
+}
```
To make this example simple, let's remove the read from the CLI part and just return a set of IDs from this function. In a code comment, mention that we are sending a dummy ID set here, but it can be used to fetch IDs from anywhere. Might have to change the description a bit as well.

```suggestion
function getLeadsIDsfromCLI(string typeOfLeads) returns string[] {
    // This function is used to get the leads IDs from the user input.
    // In a real-world scenario, you can get the leads IDs from a database or any other source.

    string[] leadsIDs = [];
    string leadID;
    io:println("[*] Input ", typeOfLeads, " lead IDs");
    while (true) {
        io:print("[*] Enter a lead ID (or Press Enter to finish): ");
        leadID = io:readln();
        if (leadID == "") {
            break;
        }
        leadsIDs.push(leadID);
    }
    return leadsIDs;
}
```

sajitha-tj replied:
Ok, did that. [bdf091a](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/bdf091a0a9b82ddc6880244d9c229650fa369357)

ThisaruGuruge commented on examples/leads_tracker/main.bal:
```diff
@@ -0,0 +1,111 @@
+
+function addRecordsToList(string listId, string[] payload, string typeOfLeads) returns error? {
+    hubspotcrmLists:MembershipsUpdateResponse response = check hubspotClient->putListidMembershipsAdd_add(listId, payload);
+    if (response.recordsIdsAdded !is ()) {
+        io:println("[+] Leads added to the " + typeOfLeads + " Leads List: ", response.recordsIdsAdded);
+    }
+    if (response.recordIdsMissing !is ()) {
+        io:println("[!] Leads not added to the " + typeOfLeads + " Leads List: ", response.recordIdsMissing);
+    }
```
Please check whether there are more places where we used parenthesis around the `if` conditions, and remove them.

```suggestion
    if response.recordsIdsAdded !is () {
        io:println("[+] Leads added to the " + typeOfLeads + " Leads List: ", response.recordsIdsAdded);
    }
    if response.recordIdsMissing !is () {
        io:println("[!] Leads not added to the " + typeOfLeads + " Leads List: ", response.recordIdsMissing);
    }
```

sajitha-tj replied:
Done. [a1016b3](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.lists/pull/2/commits/a1016b385e9232c7a5680b5ad7ec346104867022)

ThisaruGuruge reviewed (commented):
Good job overall!

sajitha-tj committed:
Update README.md

Co-authored-by: Thisaru Guruge <thisaru@wso2.com>

sajitha-tj committed:
Merge branch 'main' of https://github.com/sajitha-tj/module-ballerinax-hubspot.crm.lists

sajitha-tj committed:
Update documentations

sajitha-tj committed:
Update ballerina/tests/tests.bal

Co-authored-by: Thisaru Guruge <thisaru@wso2.com>

sajitha-tj committed:
Merge branch 'main' of https://github.com/sajitha-tj/module-ballerinax-hubspot.crm.lists

sajitha-tj committed:
Remove parenthesis from if statements

sajitha-tj committed:
Update example code as per to PR reviews

sajitha-tj committed:
Update leads_tracker example code

Few changes were added as suggested on the PR review. The function used to get the list of lead IDs was simplified to return a list of dummy IDs. In a real world scenario, a database fetch or any other method can be implemented to fetch the relevent IDs.

sajitha-tj committed:
Update documents of examples

sajitha-tj committed:
Update documentations

sajitha-tj committed:
Correct grammar and spelling mistakes in the docs

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
Add hubspot icon

sajitha-tj committed:
Update tests to work with mock server

Since there are only a few endpoints at the mock server, some tests fail when they are executed at the build stage. This update prevents execution of these tests, when the mock server is used.

sajitha-tj committed:
Update Ballerina.toml to add hubspot icon

sajitha-tj committed:
[Automated] Update the toml files

sajitha-tj committed:
Update mock_service.bal file

Previously the mock service started even with live tests. This change will prevent execution of mock server when the live hubspot server is used for testing.

sajitha-tj committed:
Update the license

sajitha-tj committed:
Do minor changes on docs

sajitha-tj committed:
Fix tests with mock server

sajitha-tj committed:
[Automated] Update the toml files
